doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

```diff
@@ -0,0 +1,302 @@
+# Graceful Reconfiguration of Managed Clusters
+
+Associated issues/prs:
+[#20010](https://github.com/MaterializeInc/materialize/issues/20010)
+[#26401](https://github.com/MaterializeInc/materialize/pull/26401)
+
+## The Problem
+Reconfiguring a managed cluster via `ALTER CLUSTER...` leads to downtime. This
+is due to our cluster sequencer first removing the old replicas and then adding new
+replicas that match the new configuration. The duration of downtime can be seen
+as extending through the full period it takes to rehydrate the new replicas.
+
+## Success Criteria
+A mechanism should be provided that allows users to alter a managed cluster,
+delaying the deletion of old replicas while new replicas are spun up and
+hydrated.
+
+## Scope
+### V1 Scope
+ - Provide a timeout-based reconfiguration cleanup scheduling.
+ - The DDL will block for the duration of the reconfiguration.
+ - The action will be cancellable by canceling the query.
+
+### V2 Scope
+ - Provide a Hydration-based reconfiguration cleanup scheduling.
+ - The DDL will be able to run in the background.
+ - The action will be cancelable or able to be preemptively finalized.
+
+## Out of Scope
+ - Interactions with proposed schedules, ex: auto-scaling/auto-quiescing
+
+## Solution Proposal
+
+### Summary:
+This feature will introduce new SQL in `ALTER CLUSTER` which will create new
+replicas matching the alter statement; however, unlike existing alter mechanisms,
+we will delay the cleanup of old replicas until the provided conditions are met.
+
+The SQL will need to define a type of check to be performed, parameters for that
+check, and whether or not to background the operation.
+
+Suggested syntax:
+
+```sql
+ALTER CLUSTER c1 SET (SIZE 'small') WITH ( 
+  WAIT = FOR <interval>,
+  BACKGROUND = {true|false},
+)
+```
+
+```sql
+ALTER CLUSTER c1 SET (SIZE 'small') WITH ( 
+  WAIT = UNTIL CAUGHT UP (
+    TIMEOUT = <interval>,
+    ON TIMEOUT = {CONTINUE | ABORT}
+  )
+  BACKGROUND = {true|false},
+)
+```
+
+The two types of checks, in agreement with the V1 and V2 scopes respectively, are: 
+ - `FOR` which waits for a provided duration before failing and aborting.
+ - `UNTIL CAUGHT UP` which will wait for a "catch up" mechanism to return true
+   or for a timeout to be met. It will roll back or forward depending on the value
+   provided to `ON TIMEOUT`.
+
+The SQL will kick off some action in `sequence_alter_cluster_managed_to_managed`
+which will result in new reconfiguration replicas being created, new entries
+placed `mz_internal.mz_pending_cluster_replicas` and a `ReconfigureCluster`
+task being created which will ensure the reconfiguration finishes.
+
+To keep track of ongoing reconfigurations we will introduce a new
+`mz_internal.mz_pending_cluster_replicas` table which holds the state of
+reconfiguration replicas. This table will include all configuration parameters
```

Comment:
yeah, this has been removed, but should have been clarified.. It would only cover an internal and recoverable failure such as envd crash in the v2 scenario.